build: implement a Commit Queue in Actions (#34112)
nodejs/node: doc/guides/commit-queue.md

# Commit Queue

> tl;dr: You can land Pull Requests by adding the `commit-queue` label to it.

> > This feature is still experimental.

Reviewer: Why is this a quote though?

Author: Just to show with different formatting, could also be bold or italic

Reviewer: I think it's fine to just have this at the beginning without specific formatting.
As for the `This feature is still experimental.` I'd go with the formatting we have for docs.
So:

Pull Requests can be landed automatically by adding the `commit-queue` label to it.

> Stability: 1 - Experimental

> 

Reviewer: We should probably state that adding a label `implies` the `certify safe` by the collaborator check the same way it is explicitly done on CI.

Author: I think this makes sense for the other PR, but not for this one. We don't have a "certify safety" when landing PRs in our current process

Reviewer: I guess it would be better to just make sure we check every requirement for PR landing then.

Author: We do the same checks collaborators already do today by calling `git node land`. Can you think of anything we might not be checking at the moment?

Commit Queue is an experimental feature for the project which simplifies the
landing process by automating it via GitHub Actions. With it, Collaborators are
able to land Pull Requests by adding the `commit-queue` label to a PR. All
checks will run via node-core-utils, and if the Pull Request is ready to land,
the Action will rebase it and push to master.

This document gives an overview on how the Commit Queue works, as well as
implementation details, reasoning for design choices, and current limitations.

## Overview

From a high-level, the Commit Queue works as follow:

1. Collaborators will add `commit-queue` label to Pull Reuqests ready to land

> 2. Every five minutes the queue will do the followith with each Pull Request with the label:

Reviewer 2: ```suggestion
2. Every five minutes the queue will do the following for each Pull Request with the label:
```

  1. Check if the PR also has a `request-ci` label (if it has, skip this PR since it's pending a CI run)
  2. Check if the last Jenkins CI is finished running (if it is not, skip this PR)
  3. Remove the `commit-queue` label
  4. Run `git node land <pr>`
  5. If it fails:
    1. Abort `git node land` session
    2. Add `commit-queue-failed` label to the PR
    3. Leave a comment on the PR with the output from `git node land`
    4. Skip next steps, go to next PR in the queue
  6. If it succeeds:
    1. Push the changes to nodejs/node
    2. Leave a comment on the PR with `Landed in ...`
    3. Close the PR
    4. Go to next PR in the queue

## Current Limitations

The Commit Queue feature is still in early stages, and as such it might not
work for more complex Pull Requests. These are the currently known limitations
of the commit queue:

1. The Pull Request must have only one commit

> 2. A CI must've ran and succeded since the last change on the PR

Reviewer 2: ```suggestion
2. A CI must've ran and succeeded since the last change on the PR
```

`octokit/graphql-action` is used to fetch all Pull Requests with the
`commit-queue` label. The output is a JSON payload, so `jq` is used to turn
that into a list of PR ids we can pass as arguments to
[`commit-queue.sh`](./tools/actions/commit-queue.sh).

> The personal token only needs permission for public repositories and to read
> profiles, we can use the GITHUB_TOKEN for write operations. Jenkins token is
> required to check CI status.

`commit-queue.sh` receives the following positional arguments:

1. The repository owner
2. The repository name
3. The Action GITHUB_TOKEN
4. Every positional argument starting at this one will be a Pull Reuqest ID of
    a Pull Request with commit-queue set.

The script will iterate over the pull requests. `ncu-ci` is used to check if
the last CI is still pending, and calls to the GitHub API are used to check if
the PR is waiting for CI to start (`request-ci` label). The PR is skipped if CI
is pending. No other CI validation is done here since `git node land` will fail
if the last CI failed.

The script removes the `commit-queue` label. It then runs `git node land`,
forwarding stdout and stderr to a file. If any errors happens, 

> `git node land --abort` is ran, and then a `commit-queue-failed` label is added

Reviewer 2: ```suggestion
`git node land --abort` is run, and then a `commit-queue-failed` label is added
```

2. Every five minutes the queue will do the following for each Pull Request
   with the label:
   1. Check if the PR also has a `request-ci` label (if it has, skip this PR
      since it's pending a CI run)
   2. Check if the last Jenkins CI is finished running (if it is not, skip this
      PR)
   3. Remove the `commit-queue` label
   4. Run `git node land <pr>`
   5. If it fails:
      1. Abort `git node land` session
      2. Add `commit-queue-failed` label to the PR
      3. Leave a comment on the PR with the output from `git node land`
      4. Skip next steps, go to next PR in the queue
   6. If it succeeds:
      1. Push the changes to nodejs/node
      2. Leave a comment on the PR with `Landed in ...`
      3. Close the PR
      4. Go to next PR in the queue

## Current Limitations

The Commit Queue feature is still in early stages, and as such it might not
work for more complex Pull Requests. These are the currently known limitations
of the commit queue:

1. The Pull Request must have only one commit
2. A CI must've ran and succeeded since the last change on the PR
3. A Collaborator must have approved the PR since the last change

> 4. Only Jenkins CI is checked (Actions, V8 CI and CITGM are ignored)

Reviewer 3: I think Actions are checked now?

Author: Not yet, I have a PR open for that
